// unirand/lib/methods/bates.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/**
 * Bates distribution
 * https://en.wikipedia.org/wiki/Bates_distribution
 * @param a: number (any value)
 * @param b: number (any value), b > a
 * @param n: integer, n >= 1
 * @returns random number with Bates distribution
 * Created by Alexey S. Kiselev
 */

var Uniform = require('./uniform');

var Bates = function () {
    function Bates(n) {
        var a = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0;
        var b = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 1;

        _classCallCheck(this, Bates);

        this.n = Math.floor(n);
        this.a = Math.min(a, b);
        this.b = Math.max(a, b);
        this._uniform = new Uniform(this.a, this.b);
    }

    /**
     * Generates a random number
     * @returns a Bates distributed number
     */


    _createClass(Bates, [{
        key: 'random',
        value: function random() {
            var m = 0,
                random_number = 0,
                random = this._uniform.distribution(this.n);
            for (var k = 0; k < this.n; k += 1) {
                m += 1;
                random_number += (random[k] - random_number) / m;
            }
            return random_number;
        }

        /**
         * Generates next seeded random number
         * @returns {number}
         */

    }, {
        key: 'next',
        value: function next() {
            var m = 0,
                random_number = 0,
                random = void 0;
            for (var k = 0; k < this.n; k += 1) {
                m += 1;
                random = this._uniform.next();
                random_number += (random - random_number) / m;
            }
            return random_number;
        }

        /**
         * Generates random distribution
         * @returns an array with Bates distributed numbers
         */

    }, {
        key: 'distribution',
        value: function distribution(n) {
            var batesArray = [],
                random = this._uniform.distribution(n * this.n),
                m = void 0,
                random_number = void 0;
            for (var i = 0; i < n; i += 1) {
                m = 0;
                random_number = 0;
                for (var k = 0; k < this.n; k += 1) {
                    m += 1;
                    random_number += (random[i * this.n + k] - random_number) / m;
                }
                batesArray[i] = random_number;
            }
            return batesArray;
        }
    }, {
        key: 'isError',
        value: function isError() {
            if (!this.a && this.a !== 0) {
                return { error: 'Bates distribution: you should point "a" numerical value' };
            }

            if (!this.b && this.b !== 0) {
                return { error: 'Bates distribution: you should point "b" numerical value' };
            }

            if (!this.n || this.n < 1) {
                return { error: 'Bates distribution: you should point "n" positive integer value' };
            }

            if (this.a === this.b) {
                return { error: 'Bates distribution: Parameter "b" must be greater then "a"' };
            }
            return { error: false };
        }

        /**
         * Refresh method
         * @param newA: number - new parameter "a"
         * @param newB: number - new parameter "b"
         * @param newN: number - new parameter "n"
         * This method does not return values
         */

    }, {
        key: 'refresh',
        value: function refresh(newN) {
            var newA = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0;
            var newB = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 1;

            this.n = Math.floor(newN);
            this.a = Math.min(newA, newB);
            this.b = Math.max(newA, newB);
        }

        /**
         * Class .toString method
         * @returns {string}
         */

    }, {
        key: 'toString',
        value: function toString() {
            var info = ['Bates Distribution', 'Usage: unirand.bates(' + this.n + ', ' + this.a + ', ' + this.b + ').random()'];
            return info.join('\n');
        }

        /**
         * Mean value
         * Information only
         * For calculating real mean value use analyzer
         */

    }, {
        key: 'mean',
        get: function get() {
            return (this.a + this.b) / 2;
        }

        /**
         * Variance value
         * Information only
         * For calculating real variance value use analyzer
         */

    }, {
        key: 'variance',
        get: function get() {
            return Math.pow(this.b - this.a, 2) / (12 * this.n);
        }

        /**
         * Skewness value
         * Information only
         * For calculating real skewness value use analyzer
         */

    }, {
        key: 'skewness',
        get: function get() {
            return 0;
        }

        /**
         * Kurtosis value
         * Information only
         */

    }, {
        key: 'kurtosis',
        get: function get() {
            return -1.2 / this.n;
        }

        /**
         * All parameters of distribution in one object
         * Information only
         */

    }, {
        key: 'parameters',
        get: function get() {
            return {
                mean: this.mean,
                variance: this.variance,
                skewness: this.skewness,
                kurtosis: this.kurtosis
            };
        }
    }]);

    return Bates;
}();

module.exports = Bates;
