// unirand/lib/methods/chisquare.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/**
 * Chi Square Distribution
 * This is continuous distribution
 * https://en.wikipedia.org/wiki/Chi-squared_distribution
 * @param k: number - degrees of freedom k > 0
 * @returns Chi Square Distributed value
 * Created by Alexey S. Kiselev
 */

var Normal = require('./normal');
var Utils = require('../utils/utils');

var ChiSquare = function () {
    function ChiSquare(k) {
        _classCallCheck(this, ChiSquare);

        this.degrees = Number(k);
        this.normal = new Normal(0, 1);
    }

    /**
     * Generates a random number
     * @returns a Chi Square distributed number
     */


    _createClass(ChiSquare, [{
        key: 'random',
        value: function random() {
            var res = 0,
                random = this.normal.distribution(this.degrees);
            for (var i = 0; i < this.degrees; i += 1) {
                res += Math.pow(random[i], 2);
            }
            return res;
        }

        /**
         * Generates next seeded random number
         * @returns {number}
         */

    }, {
        key: 'next',
        value: function next() {
            var res = 0,
                random = void 0;
            for (var i = 0; i < this.degrees; i += 1) {
                random = this.normal.next();
                res += Math.pow(random, 2);
            }
            return res;
        }

        /**
         * Generates Chi Square distributed numbers
         * @param n: number - Number of elements in resulting array, n > 0
         * @returns Array<number> - Chi Square distributed numbers
         */

    }, {
        key: 'distribution',
        value: function distribution(n) {
            var chiSquareArray = [],
                res = void 0,
                random = this.normal.distribution(n * this.degrees);
            for (var i = 0; i < n; i += 1) {
                res = 0;
                for (var j = 0; j < this.degrees; j += 1) {
                    res += Math.pow(random[i * this.degrees + j], 2);
                }
                chiSquareArray[i] = res;
            }
            return chiSquareArray;
        }

        /**
         * Error handling
         * Parameter "k" must be positive
         * @returns {boolean}
         */

    }, {
        key: 'isError',
        value: function isError() {
            if (!this.degrees) {
                return { error: 'Chi Square distribution: you should point parameter "k" positive numerical value' };
            }
            if (this.degrees <= 0) {
                return { error: 'Chi Square distribution: parameter "k" must be positive integer' };
            }
            return { error: false };
        }

        /**
         * Refresh method
         * @param newK: number - new parameter "k"
         * This method does not return values
         */

    }, {
        key: 'refresh',
        value: function refresh(newK) {
            this.degrees = Number(newK);
        }

        /**
         * Class .toString method
         * @returns {string}
         */

    }, {
        key: 'toString',
        value: function toString() {
            var info = ['Chi Square Distribution', 'Usage: unirand.chisquare(' + this.degrees + ').random()'];
            return info.join('\n');
        }

        /**
         * Mean value
         * Information only
         * For calculating real mean value use analyzer
         */

    }, {
        key: 'mean',
        get: function get() {
            return this.degrees;
        }

        /**
         * Median value (approximate value)
         * Information only
         * For calculating real median value use analyzer
         */

    }, {
        key: 'median',
        get: function get() {
            return this.degrees * Math.pow(1 - 2 / (9 * this.degrees), 3);
        }

        /**
         * Mode value
         * Information only
         * For calculating real mode value use analyzer
         */

    }, {
        key: 'mode',
        get: function get() {
            return Math.max(this.degrees - 2, 0);
        }

        /**
         * Variance value
         * Information only
         * For calculating real variance value use analyzer
         */

    }, {
        key: 'variance',
        get: function get() {
            return 2 * this.degrees;
        }

        /**
         * Skewness value
         * Information only
         * For calculating real skewness value use analyzer
         */

    }, {
        key: 'skewness',
        get: function get() {
            return Math.sqrt(8 / this.degrees);
        }

        /**
         * Kurtosis value
         * Information only
         * For calculating real kurtosis value use analyzer
         */

    }, {
        key: 'kurtosis',
        get: function get() {
            return 12 / this.degrees;
        }

        /**
         * Entropy value
         * Information only
         * For calculating real entropy value use analyzer
         */

    }, {
        key: 'entropy',
        get: function get() {
            return 0.5 * this.degrees + Math.log(2 * Utils.gamma(this.degrees / 2)) + (1 - 0.5 * this.degrees) * Utils.digamma(this.degrees / 2);
        }

        /**
         * All parameters of distribution in one object
         * Information only
         */

    }, {
        key: 'parameters',
        get: function get() {
            return {
                mean: this.mean,
                median: this.median,
                mode: this.mode,
                variance: this.variance,
                skewness: this.skewness,
                entropy: this.entropy,
                kurtosis: this.kurtosis
            };
        }
    }]);

    return ChiSquare;
}();

module.exports = ChiSquare;
