// unirand/lib/methods/extremevalue.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

/**
 * Extreme (Gumbel-type) Value distribution
 * This is continuous distribution
 * https://en.wikipedia.org/wiki/Generalized_extreme_value_distribution
 * @param mu <number> - location, any value
 * @param sigma <number> - scale, sigma > 0
 * @returns Extreme Value distributed numbers
 * Created by Alexey S. Kiselev on 27.11.2017.
 */

var _prngProxy = require('../prng/prngProxy');

var _prngProxy2 = _interopRequireDefault(_prngProxy);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var ExtremeValue = function () {
    function ExtremeValue(mu, sigma) {
        _classCallCheck(this, ExtremeValue);

        this.mu = Number(mu);
        this.sigma = Number(sigma);
    }

    /**
     * Generates a random number
     * @returns a extreme value distributed number
     */


    _createClass(ExtremeValue, [{
        key: 'random',
        value: function random() {
            return this._random(_prngProxy2.default.random());
        }

        /**
         * Generates next seeded random number
         * @returns {number}
         */

    }, {
        key: 'next',
        value: function next() {
            return this._random(_prngProxy2.default.next());
        }
    }, {
        key: '_random',
        value: function _random(u) {
            return this.mu - this.sigma * Math.log(-Math.log(u));
        }

        /**
         * Generates extreme value distributed numbers
         * @param n: number - Number of elements in resulting array, n > 0
         * @returns Array<number> - extreme value distributed numbers
         */

    }, {
        key: 'distribution',
        value: function distribution(n) {
            var random = _prngProxy2.default.random(n);
            var extremeValueArray = [];
            for (var i = 0; i < n; i += 1) {
                extremeValueArray[i] = this._random(random[i]);
            }
            return extremeValueArray;
        }

        /**
         * Error handling
         * @returns {boolean}
         */

    }, {
        key: 'isError',
        value: function isError() {
            if (!this.mu && this.mu !== 0 || !this.sigma) {
                return { error: 'Extreme Value (Gumbel type) distribution: you should point parameters "mu" and "sigma" with numerical values' };
            }
            if (this.sigma <= 0) {
                return { error: 'Extreme Value (Gumbel type) distribution: parameter "sigma" must be a positive number' };
            }
            return { error: false };
        }

        /**
         * Refresh method
         * @param newMu: number - new parameter "mu"
         * @param newSigma: number - new parameter "sigma"
         * This method does not return values
         */

    }, {
        key: 'refresh',
        value: function refresh(newMu, newSigma) {
            this.mu = Number(newMu);
            this.sigma = Number(newSigma);
        }

        /**
         * Class .toString method
         * @returns {string}
         */

    }, {
        key: 'toString',
        value: function toString() {
            var info = ['Extreme Value (Gumbel type) Distribution', 'Usage: unirand.extremevalue(' + this.mu + ', ' + this.sigma + ').random()'];
            return info.join('\n');
        }

        /**
         * Mean value
         * Information only
         * Calculate this value using Euler constant
         * For calculating real mean value use analyzer
         */

    }, {
        key: 'mean',
        get: function get() {
            return this.mu + this.sigma * 0.57721566490153286;
        }

        /**
         * Median value
         * Information only
         * For calculating real median value use analyzer
         */

    }, {
        key: 'median',
        get: function get() {
            return this.mu - this.sigma * Math.log(Math.log(2));
        }

        /**
         * Mode value - value, which appears most often
         * Information only
         * For calculating real mode value use analyzer
         */

    }, {
        key: 'mode',
        get: function get() {
            return this.mu;
        }

        /**
         * Variance value
         * Information only
         * For calculating real variance value use analyzer
         */

    }, {
        key: 'variance',
        get: function get() {
            return Math.pow(this.sigma * Math.PI, 2) / 6;
        }

        /**
         * Entropy value
         * Information only
         * For calculating real entropy value use analyzer
         */

    }, {
        key: 'entropy',
        get: function get() {
            return Math.log(this.sigma) + 1.57721566490153286;
        }

        /**
         * Skewness value
         * Information only
         * For calculating real value of skewness use analyzer
         */

    }, {
        key: 'skewness',
        get: function get() {
            var riemann = 1.202056903159594;
            return 12 * Math.sqrt(6) * riemann / Math.pow(Math.PI, 3);
        }

        /**
         * Kurtosis value
         * Information only
         * For calculating real value of kurtosis use analyzer
         */

    }, {
        key: 'kurtosis',
        get: function get() {
            return 12 / 5;
        }

        /**
         * All parameters of distribution in one object
         * Information only
         */

    }, {
        key: 'parameters',
        get: function get() {
            return {
                mean: this.mean,
                median: this.median,
                mode: this.mode,
                variance: this.variance,
                entropy: this.entropy,
                skewness: this.skewness,
                kurtosis: this.kurtosis
            };
        }
    }]);

    return ExtremeValue;
}();

module.exports = ExtremeValue;
