// unirand/lib/methods/rayleigh.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

/**
 * Rayleigh distribution
 * This is continuous distribution
 * https://en.wikipedia.org/wiki/Rayleigh_distribution
 * @param sigma <number> - scale parameter, sigma > 0
 * @returns Rayleigh distributed value
 * Created by Alexey S. Kiselev
 */

var _prngProxy = require('../prng/prngProxy');

var _prngProxy2 = _interopRequireDefault(_prngProxy);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Rayleigh = function () {
    function Rayleigh(sigma) {
        _classCallCheck(this, Rayleigh);

        this.sigma = Number(sigma);
    }

    /**
     * Generates a random number
     * @returns a Rayleigh distributed number
     */


    _createClass(Rayleigh, [{
        key: 'random',
        value: function random() {
            var epsilon = 0.00000001;
            var u = Math.min(_prngProxy2.default.random() + epsilon, 1 - epsilon);
            return this._random(u);
        }

        /**
         * Generates next seeded random number
         * @returns {number}
         */

    }, {
        key: 'next',
        value: function next() {
            var epsilon = 0.00000001;
            var u = Math.min(_prngProxy2.default.next() + epsilon, 1 - epsilon);
            return this._random(u);
        }
    }, {
        key: '_random',
        value: function _random(u) {
            return this.sigma * Math.sqrt(-2 * Math.log(u));
        }

        /**
         * Generates Rayleigh distributed numbers
         * @param n: number - Number of elements in resulting array, n > 0
         * @returns Array<number> - Rayleigh distributed numbers
         */

    }, {
        key: 'distribution',
        value: function distribution(n) {
            var rayleighArray = [],
                epsilon = 0.00000001,
                random = _prngProxy2.default.random(n),
                u = void 0;
            for (var i = 0; i < n; i += 1) {
                u = Math.min(random[i] + epsilon, 1 - epsilon);
                rayleighArray[i] = this._random(u);
            }
            return rayleighArray;
        }

        /**
         * Error handling
         * @returns {boolean}
         */

    }, {
        key: 'isError',
        value: function isError() {
            if (!this.sigma) {
                return { error: 'Rayleigh distribution: you should point "sigma" (scale) numerical value' };
            }
            if (this.sigma <= 0) {
                return { error: 'Rayleigh distribution: parameter "sigma" (scale) must be a positive value' };
            }
            return { error: false };
        }

        /**
         * Refresh method
         * @param newSigma: number - new parameter "sigma"
         * This method does not return values
         */

    }, {
        key: 'refresh',
        value: function refresh(newSigma) {
            this.sigma = Number(newSigma);
        }

        /**
         * Class .toString method
         * @returns {string}
         */

    }, {
        key: 'toString',
        value: function toString() {
            var info = ['Rayleigh Distribution', 'Usage: unirand.rayleigh(' + this.sigma + ').random()'];
            return info.join('\n');
        }

        /**
         * Mean value
         * Information only
         * For calculating real mean value use analyzer
         */

    }, {
        key: 'mean',
        get: function get() {
            return this.sigma * Math.sqrt(Math.PI / 2);
        }

        /**
         * Median value
         * Information only
         * For calculating real median value use analyzer
         */

    }, {
        key: 'median',
        get: function get() {
            return this.sigma * Math.sqrt(2 * Math.log(2));
        }

        /**
         * Mode value - value, which appears most often
         * Information only
         * For calculating real mode value use analyzer
         */

    }, {
        key: 'mode',
        get: function get() {
            return this.sigma;
        }

        /**
         * Variance value
         * Information only
         * For calculating real variance value use analyzer
         */

    }, {
        key: 'variance',
        get: function get() {
            return Math.pow(this.sigma, 2) * (4 - Math.PI) / 2;
        }

        /**
         * Entropy value
         * Information only
         * For calculating real entropy value use analyzer
         */

    }, {
        key: 'entropy',
        get: function get() {
            return 1.28860783245076643030325605 + Math.log(this.sigma / 1.4142135623730950488016887242097);
        }

        /**
         * Skewness value
         * Information only
         * For calculating real skewness value use analyzer
         */

    }, {
        key: 'skewness',
        get: function get() {
            return 2 * Math.sqrt(Math.PI) * (Math.PI - 3) / Math.pow(4 - Math.PI, 1.5);
        }

        /**
         * Kurtosis value
         * Information only
         * For calculating real kurtosis value use analyzer
         */

    }, {
        key: 'kurtosis',
        get: function get() {
            return -(6 * Math.pow(Math.PI, 2) - 24 * Math.PI + 16) / Math.pow(4 - Math.PI, 2);
        }

        /**
         * All parameters of distribution in one object
         * Information only
         */

    }, {
        key: 'parameters',
        get: function get() {
            return {
                mean: this.mean,
                median: this.median,
                mode: this.mode,
                variance: this.variance,
                entropy: this.entropy,
                skewness: this.skewness,
                kurtosis: this.kurtosis
            };
        }
    }]);

    return Rayleigh;
}();

module.exports = Rayleigh;
